### PR TITLE
add scripts to fill the database

### PR DESCRIPTION
**Pull request by amdwit**
My first backend coding here. Please let me know what you think.

This is a first attempt to write a script that fills the database with representative data. It fills the database with many users.

### CODE REVIEW

**Comment by aapeliv** [comment]
Thanks, this is a great idea. We need to fix up a bunch of slow queries, and this will help.

Is it necessary to add the 2.7 MB cities file? Where is it from? What's the license? You're pulling in Faker, and that seems to already have a bunch of geo functionality (there is [this file](https://github.com/joke2k/faker/blob/master/faker/providers/geo/__init__.py) of cities). Faker also has a name generator, why not use that? It seems faker can do a lot of what you've done here manually.

(Also, don't run `hash_password` in the loop! It's slow on purpose, to avoid bruteforcing and make any stolen database useless to crack, but running it here in the fake user generator will make this take hours to run.)

**Comment by amdwit** [comment]
I will give every user the same password. The data is from : 
https://simplemaps.com/data/world-cities
For realistic data I need users living in all bigger cities, user amounts relative to city size, I do not think I can get that with faker.
I will use faker more and remove names.

**Comment by lucaslcode** [comment]
> For realistic data I need users living in all bigger cities

Why? Could you use "fake" big cities which don't correspond to actual cities?

**Comment by amdwit** [comment]
Well, maybe we must set our goal clear then. My goal was to have a realistic set of users. Not only users, but also a realistic database state as a whole, for a reasonable effort. Using fake big cities is less realistic and a bigger effort.

File: app/backend/src/createusers.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+                print(row[0])
+                cities.append(row)
+                worldpopulation += int(float(row[9]))
+                countries.add(row[4])
+    except csv.Error as e:
+        sys.exit('file {}, line {}: {}'.format(file_name, reader.line_num, e))
+
+useramount = 1000000
+users = []
+hashedpassword = hash_password('MAGA2020')
```
Comment (aapeliv) [inline app/backend/src/createusers.py]: Please don't...

Comment (amdwit) [reply app/backend/src/createusers.py]: If you meant the password: I will change that now.

Comment (aapeliv) [reply app/backend/src/createusers.py]: Thanks :)

**Comment by kalvdans** [comment]
@amdwit: Please write a bit more descriptive commit messages, they will be part of the history of the file if we merge this branch without squashing commits. Now all the commits have the message "add scripts to fill the database".

**Comment by amdwit** [comment]
Ok, clear. I will do so. Do we use a certain workflow guide, which describes for instance rules for commit comments?

**Comment by kalvdans** [comment]
> Do we use a certain workflow guide, which describes for instance rules for commit comments?

Besides the branch naming described [here](https://github.com/Couchers-org/couchers/blob/develop/docs/contributing.md#user-content-the-development-process) we have no further guidelines. A general guideline for commit messages are to keep them within 50 characters using an imperative form, see https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html .

Anyway, if you ever need to dig into the history with "git blame", you will start appeciate good commit messages :)

File: app/backend/src/createusers.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        reader.__next__()
+        for row in reader:
+            if row[9] != "":
+                cities.append(row)
+                world_population += int(float(row[9]))
+                countries.add(row[4])
+    except csv.Error as e:
+        sys.exit("file {}, line {}: {}".format(file_name, reader.line_num, e))
+
+config.check_config()
```
Comment (kalvdans) [inline app/backend/src/createusers.py]: Remove this line; it requires a ton of unnecessary environment variables to be set; only DATABASE_CONNECTION_STRING is used.

File: app/backend/src/createusers.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                countries_lived="|".join(random.choices(list(countries), k=8)),
+                hosting_status=hostingstatus2sql[random.randint(1, 5)],
+            )
+            session.add(new_user)
+        session.commit()
+
+
+def generate_dummy_data():
+    with multiprocessing.Pool(os.cpu_count() * 2) as pool:
+        pool.map(generate_city_users, cities)
```
Comment (kalvdans) [inline app/backend/src/createusers.py]: The exceptions generated by psycopg2 seems to not be pickleable so they can't be propagated through multiprocessing. I don't know what the best way to handle this is, just saying there can be an issue.
```
Traceback (most recent call last):
  File "/home/chn/couchers/couchers/app/backend/src/createusers.py", line 89, in <module>
    generate_dummy_data()
  File "/home/chn/couchers/couchers/app/backend/src/createusers.py", line 86, in generate_dummy_data
    pool.map(generate_city_users, cities)
  File "/usr/lib/python3.9/multiprocessing/pool.py", line 364, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/lib/python3.9/multiprocessing/pool.py", line 771, in get
    raise self._value
multiprocessing.pool.MaybeEncodingError: Error sending result: '<multiprocessing.pool.ExceptionWithTraceback object at 0x7feaa4426820>'. Reason: 'TypeError("cannot pickle 'psycopg2.extensions.Binary' object")'
```

File: app/backend/src/createusers.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+fake = faker.Faker()
+
+logger = logging.getLogger(__name__)
+
+cities = []
+countries = set()
+world_population = 0
+
+# data from https://simplemaps.com/data/world-cities
+file_name = "src/data/worldcities.csv"
```
Comment (kalvdans) [inline app/backend/src/createusers.py]: You could use `os.path.dirname(__file__)` to make this path independent of the current working directory.

Great that you specify the source.

**Review by kalvdans — DISMISSED** [review]
I took your script for a spin, and found some minor nuances. Looks good to merge with or without fixing them :+1:

File: app/backend/src/createusers.py (hunk 3 above)
Comment (amdwit) [reply app/backend/src/createusers.py]: I have not yet resolved this. For now I have isolated the db calls in a separate process. In a later PR I will look at it better.

**Review by kalvdans — APPROVED** [review]
LGTM!

**Pull request merged**